fuzz-systemctl-parse-argv: avoid "leak" of bus object

keszybz opened the pull request:
Memory sanitizer would report leaked memory from --boot-load-entry=help.

Maybe we should disable all bus connections from the fuzzer? It seems not
appropriate to communicate with logind. OTOH, in a real fuzzing environment
this call should just fail, so maybe that's OK.

I think this should help with the fuzzing failures reported in CI.

keszybz commented:
I found another leak.

poettering commented:
this should probably also call static_destruct() and mac_selinux_finish().

poettering commented:
i.e. everything that systemctl.c's main() function calls at the end plus the stuff from _DEFINE_MAIN_FUNCTION() that undoes the stuff the cmdline parsers set up

keszybz commented:
> this should probably also call static_destruct() and mac_selinux_finish().

Easier said that done ;) I've been working on this for a few hours… But a patch is coming up.

keszybz commented:
> this should probably also call static_destruct() and mac_selinux_finish().

Updated. The diffstat is a bit bigger now… It essentially implements the call to `static_destruct()`. I also did some unrelated cleanup in code that I was touching anyway.

poettering commented on src/analyze/analyze.c:
```diff
@@ -92,7 +92,7 @@ static usec_t arg_base_time = USEC_INFINITY;
 STATIC_DESTRUCTOR_REGISTER(arg_dot_to_patterns, strv_freep);
 
-struct boot_times {
+typedef struct BootTimes {
```
once upon a time we had this rule that noone followed that structs specific to a source file should use "struct" and no "typedef" and be lower_case and stuff exported to other .c files should use typedefs and user CamelCase. But i guess it's a mess right now, and the rule might not be the smartest rule to have, hence looks ok to me

keszybz replied:
Oh, OK. Anyway, I don't think this is even remotely true. We should probably convert the remaining few instances without typedef too.

poettering commented on src/resolve/resolved-dns-transaction.c:
```diff
@@ -141,23 +141,23 @@ DnsTransaction* dns_transaction_free(DnsTransaction *t) {
 
-        return true;
+        return t;
 }
```
interesting model

poettering commented on src/network/networkd-util.c:
```diff
@@ -155,10 +155,6 @@ int network_config_section_new(const char *filename, unsigned line, NetworkConfi
 
-void network_config_section_free(NetworkConfigSection *cs) {
-        free(cs);
-}
```
i wonder if we have similar cases elsewhere…

keszybz replied:
Not many I think. Anyway, the compiler should mostly optimize this away when doing lto. So it's mostly a matter of saving a few lines of sources.

poettering commented on src/analyze/analyze.c:
```diff
@@ -213,10 +213,10 @@ static int compare_unit_start(const UnitTimes *a, const UnitTimes *b) {
 
-static void unit_times_free(UnitTimes *t) {
-        for (UnitTimes *p = t; p->has_data; p++)
+static UnitTimes* unit_times_free(UnitTimes *t) {
```
maybe add a comment here, that this destroy the whole frickin' array. The function is called so innocently, one might not expect that

poettering commented on src/basic/strbuf.c:
```diff
@@ -65,13 +65,13 @@ void strbuf_complete(struct strbuf *str) {
 
 /* clean up everything */
-void strbuf_cleanup(struct strbuf *str) {
+struct strbuf* strbuf_cleanup(struct strbuf *str) {
```
maybe rename strbuf_free?

poettering commented on src/busctl/busctl.c:
```diff
@@ -780,24 +780,22 @@ static int member_compare_funcp(Member * const *a, Member * const *b) {
+        return set_free_with_destructor(s, member_free);
 }
-
 DEFINE_TRIVIAL_CLEANUP_FUNC(Set*, member_set_free);
```
i guess this should use a destructor in its hash_ops...

poettering commented on src/basic/macro.h:
```diff
@@ -405,10 +405,20 @@ static inline int __coverity_check_and_return__(int condition) {
                 func(p);                                        \
         }
 
+/* When func() returns the appropriate type */
```
comment should say *NULL cast to the appropriate type*.

i.e. the NULL matters

keszybz replied:
Actually it's not NULL — in some cases it's e.g. -1. If it was always NULL, I wouldn't need the second variant of DEFINE_TRIVIAL_CLEANUP_FUNC. I'll try to improve the comment.

poettering commented on src/systemctl/systemctl-compat-shutdown.c:
```diff
@@ -132,9 +132,10 @@ int shutdown_parse_argv(int argc, char *argv[]) {
+                char **copy = strv_copy(wall);
+                if (!copy)
                         return log_oom();
+                strv_free_and_replace(arg_wall, copy);
```
maybe add a helper strv_free_and_copy(), taking inspiration from free_and_strdup()?

poettering commented:
lgtm, just some nitpicks

yuwata commented on src/analyze/analyze.c:
```diff
@@ -414,28 +411,28 @@ static int acquire_time_data(sd_bus *bus, struct unit_times **out) {
         _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
         _cleanup_(sd_bus_flush_close_unrefp) sd_bus *system_bus = NULL;
-        _cleanup_(free_host_infop) struct host_info *host;
+        _cleanup_(free_host_infop) HostInfo *host;
```
I'd like to add ` = NULL` here for `host`. Of course, it is redundant now. But one day, e.g. a code may be added before `host = new0()`.

yuwata reviewed (approved):
LGTM. One minor comment.

yuwata commented:
CI fails.
```
In file included from ../src/basic/fd-util.h:9,
                 from ../src/core/apparmor-setup.c:11:
../src/core/apparmor-setup.c: In function ‘aa_policy_cache_unrefp’:
../src/basic/macro.h:412:28: error: void value not ignored as it ought to be
  412 |                         *p = func(*p);                          \
      |                            ^
../src/core/apparmor-setup.c:19:1: note: in expansion of macro ‘DEFINE_TRIVIAL_CLEANUP_FUNC’
   19 | DEFINE_TRIVIAL_CLEANUP_FUNC(aa_policy_cache *, aa_policy_cache_unref);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/core/apparmor-setup.c: In function ‘aa_features_unrefp’:
../src/basic/macro.h:412:28: error: void value not ignored as it ought to be
  412 |                         *p = func(*p);                          \
      |                            ^
../src/core/apparmor-setup.c:20:1: note: in expansion of macro ‘DEFINE_TRIVIAL_CLEANUP_FUNC’
   20 | DEFINE_TRIVIAL_CLEANUP_FUNC(aa_features *, aa_features_unref);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

keszybz commented:
Updated:
- two commits to rename the functions that @poettering mentioned
- I fixed all the "hidden" comments (NULL initialization, apparmor failure, strbuf_cleanup rename, unit_times_freep rename)

yuwata commented:
LGTM.
